services/QuillLMS/app/services/pack_sequence_item_status_mapper.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+class PackSequenceItemStatusMapper < ApplicationService
+  attr_reader :pack_sequence_item_results, :pack_sequence_item_statuses
+
+  LOCKED = PackSequenceItem::LOCKED
+  UNLOCKED = PackSequenceItem::UNLOCKED
+
+  def initialize(pack_sequence_item_results)
+    @pack_sequence_item_results = pack_sequence_item_results
+    @pack_sequence_item_statuses = {}
+  end
+
+
+  def run
+    map_statuses
+    pack_sequence_item_statuses
+  end
+
+  private def map_statuses
+    previous_packs_unfinished = false
+    previous_id = nil
+    status = UNLOCKED
+
+    ordered_pack_sequence_item_results.each do |pack_sequence_item_result|
```

Comment:
So the `locked/unlocked` `status` of an `item` isn't stored in the DB and is live calculated on each call. I mentioned this on the RFC, but this might have performance issues long-term. It will likely be fine for an individual student, but it does make it a more complicated calculation for other uses, e.g. a teacher with 30 students needs to see which activities for their class are `unlocked` at a given moment, or a school wanted to know that, or we wanted to run some analytics queries just in `SQL`. It does have the benefit of not getting out of sync, since it's always calculated live, so there is an upside to this approach (and `state` is always tricky to deal with).

So I'm ok with trying this approach, just with the caveat that if we do hit issues, we may have to pivot to a different approach down the road.

Reply:
Given the complexity of this individual case, it did make me wonder about moving the value into the database.  There are however, various edge cases that would add even more complexity for such an implementation.  

Another option would be adding a raw SQL solution involving a lateral join, but I felt at this point that would be overkill.

That being said, I'm perfectly find with pivoting if this current approach is not performant.